// @aws-cdk/aws-efs/test/access-point.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
const ec2 = require("@aws-cdk/aws-ec2");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
let stack;
let vpc;
let fileSystem;
beforeEach(() => {
    stack = new core_1.Stack();
    vpc = new ec2.Vpc(stack, 'VPC');
    fileSystem = new lib_1.FileSystem(stack, 'EfsFileSystem', {
        vpc,
    });
});
test('addAccessPoint correctly', () => {
    // WHEN
    fileSystem.addAccessPoint('MyAccessPoint');
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::EFS::AccessPoint'));
});
test('new AccessPoint correctly', () => {
    // WHEN
    new lib_1.AccessPoint(stack, 'MyAccessPoint', {
        fileSystem,
    });
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::EFS::AccessPoint'));
});
test('import correctly', () => {
    // WHEN
    const ap = new lib_1.AccessPoint(stack, 'MyAccessPoint', {
        fileSystem,
    });
    const imported = lib_1.AccessPoint.fromAccessPointId(stack, 'ImportedAccessPoint', ap.accessPointId);
    // THEN
    expect(imported.accessPointId).toEqual(ap.accessPointId);
});
test('custom access point is created correctly', () => {
    // WHEN
    new lib_1.AccessPoint(stack, 'MyAccessPoint', {
        fileSystem,
        createAcl: {
            ownerGid: '1000',
            ownerUid: '1000',
            permissions: '755',
        },
        path: '/export/share',
        posixUser: {
            gid: '1000',
            uid: '1000',
            secondaryGids: [
                '1001',
                '1002',
            ],
        },
    });
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::EFS::AccessPoint', {
        FileSystemId: {
            Ref: 'EfsFileSystem37910666',
        },
        PosixUser: {
            Gid: '1000',
            SecondaryGids: [
                '1001',
                '1002',
            ],
            Uid: '1000',
        },
        RootDirectory: {
            CreationInfo: {
                OwnerGid: '1000',
                OwnerUid: '1000',
                Permissions: '755',
            },
            Path: '/export/share',
        },
    }));
});
