// @aws-cdk/aws-ec2/test/integ.reserved-private-subnet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/core");
const ec2 = require("../lib");
/*
 * Creates a simple vpc with a public subnet and a private reserved subnet.
 * Public subnet should be visible but a private reserved subnet only has IP space reserved.
 * No resources are provisioned in a reserved subnet.
 *
 * Stack verification steps:
 * -- aws ec2 describe-nat-gateways returns { "natGateways": []}
 */
const app = new cdk.App();
class VpcReservedPrivateSubnetStack extends cdk.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        /// !show
        // Specify no NAT gateways with a reserved private subnet
        new ec2.Vpc(this, 'VPC', {
            cidr: '10.0.0.0/16',
            subnetConfiguration: [
                {
                    name: 'ingress',
                    subnetType: ec2.SubnetType.PUBLIC,
                },
                {
                    name: 'private',
                    subnetType: ec2.SubnetType.PRIVATE,
                    reserved: true,
                },
            ],
            natGateways: 0,
        });
        /// !hide
    }
}
new VpcReservedPrivateSubnetStack(app, 'aws-cdk-ec2-vpc-endpoint');
app.synth();
