// @aws-cdk/aws-route53/test/test.hosted-zone.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'Hosted Zone': {
        'Hosted Zone constructs the ARN'(test) {
            // GIVEN
            const stack = new cdk.Stack(undefined, 'TestStack', {
                env: { account: '123456789012', region: 'us-east-1' },
            });
            const testZone = new lib_1.HostedZone(stack, 'HostedZone', {
                zoneName: 'testZone',
            });
            test.deepEqual(stack.resolve(testZone.hostedZoneArn), {
                'Fn::Join': [
                    '',
                    [
                        'arn:',
                        { Ref: 'AWS::Partition' },
                        ':route53:::hostedzone/',
                        { Ref: 'HostedZoneDB99F866' },
                    ],
                ],
            });
            test.done();
        },
    },
    'Supports tags'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const hostedZone = new lib_1.HostedZone(stack, 'HostedZone', {
            zoneName: 'test.zone',
        });
        cdk.Tag.add(hostedZone, 'zoneTag', 'inMyZone');
        // THEN
        assert_1.expect(stack).toMatch({
            Resources: {
                HostedZoneDB99F866: {
                    Type: 'AWS::Route53::HostedZone',
                    Properties: {
                        Name: 'test.zone.',
                        HostedZoneTags: [
                            {
                                Key: 'zoneTag',
                                Value: 'inMyZone',
                            },
                        ],
                    },
                },
            },
        });
        test.done();
    },
};
