// @aws-cdk/aws-lambda/test/integ.bundling.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const core_1 = require("@aws-cdk/core");
const lambda = require("../lib");
/**
 * Stack verification steps:
 * * aws cloudformation describe-stacks --stack-name cdk-integ-lambda-bundling --query Stacks[0].Outputs[0].OutputValue
 * * aws lambda invoke --function-name <output from above> response.json
 * * cat response.json
 * The last command should show '200'
 */
class TestStack extends core_1.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        const assetPath = path.join(__dirname, 'python-lambda-handler');
        const fn = new lambda.Function(this, 'Function', {
            code: lambda.Code.fromAsset(assetPath, {
                bundling: {
                    image: lambda.Runtime.PYTHON_3_6.bundlingDockerImage,
                    command: [
                        'bash', '-c', [
                            'rsync -r . /asset-output',
                            'cd /asset-output',
                            'pip install -r requirements.txt -t .',
                        ].join(' && '),
                    ],
                },
            }),
            runtime: lambda.Runtime.PYTHON_3_6,
            handler: 'index.handler',
        });
        new core_1.CfnOutput(this, 'FunctionArn', {
            value: fn.functionArn,
        });
    }
}
const app = new core_1.App();
new TestStack(app, 'cdk-integ-lambda-bundling');
app.synth();
