// @aws-cdk/aws-route53-targets/test/cloudfront-target.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const assert_1 = require("@aws-cdk/assert");
const cloudfront = require("@aws-cdk/aws-cloudfront");
const route53 = require("@aws-cdk/aws-route53");
const s3 = require("@aws-cdk/aws-s3");
const core_1 = require("@aws-cdk/core");
const targets = require("../lib");
test('use CloudFrontTarget partition hosted zone id mapping', () => {
    // GIVEN
    const stack = new core_1.Stack();
    // WHEN
    targets.CloudFrontTarget.getHostedZoneId(stack);
    // THEN
    expect(assert_1.SynthUtils.toCloudFormation(stack)).toEqual({
        Mappings: {
            AWSCloudFrontPartitionHostedZoneIdMap: {
                'aws': {
                    zoneId: 'Z2FDTNDATAQYW2',
                },
                'aws-cn': {
                    zoneId: 'Z3RFFRIM2A3IF5',
                },
            },
        },
    });
});
test('use CloudFrontTarget hosted zone id mappings in nested stacks', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const nestedStackA = new core_1.NestedStack(stack, 'nestedStackA');
    const nestedStackB = new core_1.NestedStack(stack, 'nestedStackB');
    // WHEN
    targets.CloudFrontTarget.getHostedZoneId(nestedStackA);
    targets.CloudFrontTarget.getHostedZoneId(nestedStackB);
    // THEN
    for (let nestedStack of [nestedStackA, nestedStackB]) {
        expect(assert_1.SynthUtils.toCloudFormation(nestedStack)).toEqual({
            Mappings: {
                AWSCloudFrontPartitionHostedZoneIdMap: {
                    'aws': {
                        zoneId: 'Z2FDTNDATAQYW2',
                    },
                    'aws-cn': {
                        zoneId: 'Z3RFFRIM2A3IF5',
                    },
                },
            },
        });
    }
});
test('use CloudFront as record target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const sourceBucket = new s3.Bucket(stack, 'Bucket');
    const distribution = new cloudfront.CloudFrontWebDistribution(stack, 'MyDistribution', {
        originConfigs: [
            {
                s3OriginSource: {
                    s3BucketSource: sourceBucket,
                },
                behaviors: [{ isDefaultBehavior: true }],
            },
        ],
    });
    // WHEN
    const zone = new route53.PublicHostedZone(stack, 'HostedZone', { zoneName: 'test.public' });
    new route53.ARecord(zone, 'Alias', {
        zone,
        recordName: '_foo',
        target: route53.RecordTarget.fromAlias(new targets.CloudFrontTarget(distribution)),
    });
    // THEN
    expect(stack).toHaveResource('AWS::Route53::RecordSet', {
        AliasTarget: {
            DNSName: { 'Fn::GetAtt': ['MyDistributionCFDistributionDE147309', 'DomainName'] },
            HostedZoneId: {
                'Fn::FindInMap': [
                    'AWSCloudFrontPartitionHostedZoneIdMap',
                    {
                        Ref: 'AWS::Partition',
                    },
                    'zoneId',
                ],
            },
        },
    });
});
