// @aws-cdk/aws-route53-targets/test/integ.interface-vpc-endpoint-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ec2 = require("@aws-cdk/aws-ec2");
const route53 = require("@aws-cdk/aws-route53");
const cdk = require("@aws-cdk/core");
const targets = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-interface-vpc-endpoint', {
    env: {
        region: 'us-west-2',
    },
});
const vpc = new ec2.Vpc(stack, 'VPC');
const interfaceVpcEndpoint = new ec2.InterfaceVpcEndpoint(stack, 'InterfaceEndpoint', {
    vpc,
    service: {
        name: 'com.amazonaws.us-west-2.sms',
        port: 80,
    },
    privateDnsEnabled: false,
    subnets: {
        subnetType: ec2.SubnetType.PRIVATE,
    },
});
const zone = new route53.PrivateHostedZone(stack, 'PrivateZone', {
    vpc,
    zoneName: 'test.aws.cdk.com',
});
new route53.ARecord(stack, 'AliasEndpointRecord', {
    zone,
    recordName: 'foo',
    target: route53.RecordTarget.fromAlias(new targets.InterfaceVpcEndpointTarget(interfaceVpcEndpoint)),
});
app.synth();
