// @aws-cdk/aws-ssm/test/test.parameter-store-string.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const ssm = require("../lib");
module.exports = {
    'can reference SSMPS string - specific version'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const ref = ssm.StringParameter.fromStringParameterAttributes(stack, 'Ref', {
            parameterName: '/some/key',
            version: 123,
        });
        // THEN
        test.equal(stack.resolve(ref.stringValue), '{{resolve:ssm:/some/key:123}}');
        test.done();
    },
    'can reference SSMPS string - latest version'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const ref = ssm.StringParameter.fromStringParameterAttributes(stack, 'Ref', {
            parameterName: '/some/key',
        });
        // THEN
        assert_1.expect(stack).toMatch({
            Parameters: {
                RefParameter: {
                    Type: 'AWS::SSM::Parameter::Value<String>',
                    Default: '/some/key',
                },
            },
        });
        test.deepEqual(stack.resolve(ref.stringValue), { Ref: 'RefParameter' });
        test.done();
    },
    'can reference SSMPS secure string'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const ref = ssm.StringParameter.fromSecureStringParameterAttributes(stack, 'Ref', {
            parameterName: '/some/key',
            version: 123,
        }).stringValue;
        // THEN
        test.equal(stack.resolve(ref), '{{resolve:ssm-secure:/some/key:123}}');
        test.done();
    },
    'empty parameterName will throw'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        test.throws(() => {
            ssm.StringParameter.fromStringParameterAttributes(stack, 'Ref', {
                parameterName: '',
            });
        }, /parameterName cannot be an empty string/);
        test.done();
    },
};
