// @aws-cdk/aws-cognito/test/integ.user-pool-idp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
/*
 * Stack verification steps
 * * Visit the URL provided by stack output 'SignInLink' in a browser, and verify the 'Login with Amazon' link shows up.
 * * If you plug in valid 'Login with Amazon' credentials, the federated log in should work.
 */
const app = new core_1.App();
const stack = new core_1.Stack(app, 'integ-user-pool-idp');
const userpool = new lib_1.UserPool(stack, 'pool');
new lib_1.UserPoolIdentityProviderAmazon(stack, 'amazon', {
    userPool: userpool,
    clientId: 'amzn-client-id',
    clientSecret: 'amzn-client-secret',
    attributeMapping: {
        givenName: lib_1.ProviderAttribute.AMAZON_NAME,
        email: lib_1.ProviderAttribute.AMAZON_EMAIL,
        custom: {
            userId: lib_1.ProviderAttribute.AMAZON_USER_ID,
        },
    },
});
const client = userpool.addClient('client');
const domain = userpool.addDomain('domain', {
    cognitoDomain: {
        domainPrefix: 'nija-test-pool',
    },
});
new core_1.CfnOutput(stack, 'SignInLink', {
    value: domain.signInUrl(client, {
        redirectUri: 'https://example.com',
    }),
});
