// @aws-cdk/aws-logs/test/test.metricfilter.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const aws_cloudwatch_1 = require("@aws-cdk/aws-cloudwatch");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'trivial instantiation'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        const logGroup = new lib_1.LogGroup(stack, 'LogGroup');
        // WHEN
        new lib_1.MetricFilter(stack, 'Subscription', {
            logGroup,
            metricNamespace: 'AWS/Test',
            metricName: 'Latency',
            metricValue: '$.latency',
            filterPattern: lib_1.FilterPattern.exists('$.latency'),
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::Logs::MetricFilter', {
            MetricTransformations: [{
                    MetricNamespace: 'AWS/Test',
                    MetricName: 'Latency',
                    MetricValue: '$.latency',
                }],
            FilterPattern: '{ $.latency = "*" }',
            LogGroupName: { Ref: 'LogGroupF5B46931' },
        }));
        test.done();
    },
    'metric filter exposes metric'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        const logGroup = new lib_1.LogGroup(stack, 'LogGroup');
        // WHEN
        const mf = new lib_1.MetricFilter(stack, 'Subscription', {
            logGroup,
            metricNamespace: 'AWS/Test',
            metricName: 'Latency',
            metricValue: '$.latency',
            filterPattern: lib_1.FilterPattern.exists('$.latency'),
        });
        const metric = mf.metric();
        // THEN
        test.deepEqual(metric, new aws_cloudwatch_1.Metric({
            metricName: 'Latency',
            namespace: 'AWS/Test',
            statistic: 'avg',
        }));
        test.done();
    },
    'metric filter exposes metric with custom statistic'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        const logGroup = new lib_1.LogGroup(stack, 'LogGroup');
        // WHEN
        const mf = new lib_1.MetricFilter(stack, 'Subscription', {
            logGroup,
            metricNamespace: 'AWS/Test',
            metricName: 'Latency',
            metricValue: '$.latency',
            filterPattern: lib_1.FilterPattern.exists('$.latency'),
        });
        const metric = mf.metric({ statistic: 'maximum' });
        // THEN
        test.deepEqual(metric, new aws_cloudwatch_1.Metric({
            metricName: 'Latency',
            namespace: 'AWS/Test',
            statistic: 'maximum',
        }));
        test.done();
    },
};
