// @aws-cdk/core/test/test.cfn-json.js
"use strict";
const lib_1 = require("../lib");
const cfn_json_1 = require("../lib/cfn-json");
const index_1 = require("../lib/private/cfn-utils-provider/index");
module.exports = {
    'resolves to a fn::getatt'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack = new lib_1.Stack(app, 'test');
        // WHEN
        const json = new cfn_json_1.CfnJson(stack, 'MyCfnJson', {
            value: {
                hello: 1234,
                world: { bar: 1234 },
            },
        });
        // THEN
        const template = app.synth().getStackArtifact(stack.artifactId).template;
        // input is stringified
        test.deepEqual(template.Resources.MyCfnJson248769BB.Properties.Value, '{"hello":1234,"world":{"bar":1234}}');
        // output is basically an Fn::GetAtt
        test.deepEqual(stack.resolve(json), { 'Fn::GetAtt': ['MyCfnJson248769BB', 'Value'] });
        test.done();
    },
    'tokens and intrinsics can be used freely in keys or values'(test) {
        // GIVEN
        const app = new lib_1.App();
        const stack = new lib_1.Stack(app, 'test');
        const other = new lib_1.CfnResource(stack, 'Other', { type: 'MyResource' });
        // WHEN
        new cfn_json_1.CfnJson(stack, 'MyCfnJson', {
            value: {
                [other.ref]: 1234,
                world: {
                    bar: `this is a ${lib_1.Lazy.stringValue({ produce: () => 'I am lazy' })}`,
                },
            },
        });
        // THEN
        const template = app.synth().getStackArtifact(stack.artifactId).template;
        test.deepEqual(template.Resources.MyCfnJson248769BB.Properties.Value, {
            'Fn::Join': ['', ['{"', { Ref: 'Other' }, '":1234,"world":{"bar":"this is a I am lazy"}}']],
        });
        test.done();
    },
    'JSON.stringify() will return the CFN-stringified value to avoid circular references'(test) {
        // GIVEN
        const stack = new lib_1.Stack();
        const res = new lib_1.CfnResource(stack, 'MyResource', { type: 'Foo' });
        const cfnjson = new cfn_json_1.CfnJson(stack, 'MyCfnJson', {
            value: {
                [`ref=${res.ref}`]: `this is a ${lib_1.Lazy.stringValue({ produce: () => 'I am lazy' })}`,
            },
        });
        // WHEN
        const str = JSON.stringify(cfnjson);
        // THEN
        test.ok(typeof (str) === 'string');
        test.deepEqual(stack.resolve(str), {
            'Fn::Join': ['', ['"{"ref=', { Ref: 'MyResource' }, '":"this is a I am lazy"}"']],
        });
        test.done();
    },
    async 'resource provider simply parses json and reflects back as an attribute'(test) {
        const input = { foo: 1234 };
        const response = await index_1.handler({
            ResourceType: "Custom::AWSCDKCfnJson" /* CFN_JSON */,
            ResourceProperties: {
                Value: JSON.stringify(input),
            },
        });
        test.deepEqual(input, response.Data.Value);
        test.done();
    },
};
