// @aws-cdk/core/test/test.aspect.js
"use strict";
const cxschema = require("@aws-cdk/cloud-assembly-schema");
const lib_1 = require("../lib");
const construct_compat_1 = require("../lib/construct-compat");
class MyConstruct extends construct_compat_1.Construct {
    constructor() {
        super(...arguments);
        this.visitCounter = 0;
    }
    static IsMyConstruct(x) {
        return x.visitCounter !== undefined;
    }
}
class VisitOnce {
    visit(node) {
        if (MyConstruct.IsMyConstruct(node)) {
            node.visitCounter += 1;
        }
    }
}
class MyAspect {
    visit(node) {
        node.node.addMetadata('foo', 'bar');
    }
}
module.exports = {
    'Aspects are invoked only once'(test) {
        const app = new lib_1.App();
        const root = new MyConstruct(app, 'MyConstruct');
        root.node.applyAspect(new VisitOnce());
        app.synth();
        test.deepEqual(root.visitCounter, 1);
        app.synth();
        test.deepEqual(root.visitCounter, 1);
        test.done();
    },
    'Warn if an Aspect is added via another Aspect'(test) {
        const app = new lib_1.App();
        const root = new MyConstruct(app, 'MyConstruct');
        const child = new MyConstruct(root, 'ChildConstruct');
        root.node.applyAspect({
            visit(construct) {
                construct.node.applyAspect({
                    visit(inner) {
                        inner.node.addMetadata('test', 'would-be-ignored');
                    },
                });
            },
        });
        app.synth();
        test.deepEqual(root.node.metadata[0].type, cxschema.ArtifactMetadataEntryType.WARN);
        test.deepEqual(root.node.metadata[0].data, 'We detected an Aspect was added via another Aspect, and will not be applied');
        // warning is not added to child construct
        test.equal(child.node.metadata.length, 0);
        test.done();
    },
    'Do not warn if an Aspect is added directly (not by another aspect)'(test) {
        const app = new lib_1.App();
        const root = new MyConstruct(app, 'Construct');
        const child = new MyConstruct(root, 'ChildConstruct');
        root.node.applyAspect(new MyAspect());
        app.synth();
        test.deepEqual(root.node.metadata[0].type, 'foo');
        test.deepEqual(root.node.metadata[0].data, 'bar');
        test.deepEqual(root.node.metadata[0].type, 'foo');
        test.deepEqual(child.node.metadata[0].data, 'bar');
        // no warning is added
        test.equal(root.node.metadata.length, 1);
        test.equal(child.node.metadata.length, 1);
        test.done();
    },
};
