// @aws-cdk/core/test/test.rule.js
"use strict";
const lib_1 = require("../lib");
const util_1 = require("./util");
module.exports = {
    'Rule can be used to create rules'(test) {
        const stack = new lib_1.Stack();
        const rule = new lib_1.CfnRule(stack, 'MyRule');
        rule.addAssertion(lib_1.Fn.conditionEquals('lhs', 'rhs'), 'lhs equals rhs');
        rule.addAssertion(lib_1.Fn.conditionNot(lib_1.Fn.conditionAnd(lib_1.Fn.conditionContains(['hello', 'world'], 'world'))), 'some assertion');
        test.deepEqual(util_1.toCloudFormation(stack), {
            Rules: {
                MyRule: {
                    Assertions: [
                        {
                            Assert: { 'Fn::Equals': ['lhs', 'rhs'] },
                            AssertDescription: 'lhs equals rhs',
                        },
                        {
                            Assert: { 'Fn::Not': [{ 'Fn::And': [{ 'Fn::Contains': [['hello', 'world'], 'world'] }] }] },
                            AssertDescription: 'some assertion',
                        },
                    ],
                },
            },
        });
        test.done();
    },
    'a template can contain multiple Rules'(test) {
        const stack = new lib_1.Stack();
        new lib_1.CfnRule(stack, 'Rule1');
        new lib_1.CfnRule(stack, 'Rule2');
        test.deepEqual(util_1.toCloudFormation(stack), {
            Rules: {
                Rule1: {},
                Rule2: {},
            },
        });
        test.done();
    },
};
