// aws-cdk/test/util/mock-toolkitinfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockToolkitInfo = void 0;
const api_1 = require("../../lib/api");
function mockLike() {
    return jest.fn();
}
class MockToolkitInfo extends api_1.ToolkitInfo {
    constructor(sdk, props = {}) {
        var _a, _b, _c;
        super(sdk);
        this.found = true;
        this.prepareEcrRepository = mockLike();
        this.bucketName = (_a = props.bucketName) !== null && _a !== void 0 ? _a : 'MockToolkitBucketName';
        this.bucketUrl = (_b = props.bucketUrl) !== null && _b !== void 0 ? _b : `https://${this.bucketName}.s3.amazonaws.com/`;
        this.version = (_c = props.version) !== null && _c !== void 0 ? _c : 1;
        this._bootstrapStack = props.bootstrapStack;
    }
    get bootstrapStack() {
        if (!this._bootstrapStack) {
            throw new Error('Bootstrap stack object expected but not supplied to MockToolkitInfo');
        }
        return this._bootstrapStack;
    }
    async validateVersion(expectedVersion, ssmParameterName) {
        const version = ssmParameterName !== undefined ? await this.versionFromSsmParameter(ssmParameterName) : this.version;
        if (expectedVersion > version) {
            throw new Error(`This CDK deployment requires bootstrap stack version '${expectedVersion}', found '${version}'. Please run 'cdk bootstrap' with a newer CLI version.`);
        }
    }
}
exports.MockToolkitInfo = MockToolkitInfo;
