// aws-cdk/test/util/mock-toolkitinfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockToolkitInfo = void 0;
/* eslint-disable import/order */
const api_1 = require("../../lib/api");
function mockLike() {
    return jest.fn();
}
class MockToolkitInfo extends api_1.ToolkitInfo {
    constructor(sdk, props = {}) {
        super(sdk);
        this.found = true;
        this.prepareEcrRepository = mockLike();
        this.bucketName = props.bucketName ?? 'MockToolkitBucketName';
        this.bucketUrl = props.bucketUrl ?? `https://${this.bucketName}.s3.amazonaws.com/`;
        this.version = props.version ?? 1;
        this.variant = api_1.DEFAULT_BOOTSTRAP_VARIANT;
        this._bootstrapStack = props.bootstrapStack;
    }
    get bootstrapStack() {
        if (!this._bootstrapStack) {
            throw new Error('Bootstrap stack object expected but not supplied to MockToolkitInfo');
        }
        return this._bootstrapStack;
    }
    async validateVersion(expectedVersion, ssmParameterName) {
        const version = ssmParameterName !== undefined ? await api_1.ToolkitInfo.versionFromSsmParameter(this.sdk, ssmParameterName) : this.version;
        if (expectedVersion > version) {
            throw new Error(`This CDK deployment requires bootstrap stack version '${expectedVersion}', found '${version}'. Please run 'cdk bootstrap' with a newer CLI version.`);
        }
    }
}
exports.MockToolkitInfo = MockToolkitInfo;
