// aws-cdk/lib/api/bootstrap/bootstrap-environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bootstrapEnvironment2 = exports.bootstrapEnvironment = void 0;
const path = require("path");
const serialize_1 = require("../../serialize");
const deploy_bootstrap_1 = require("./deploy-bootstrap");
const legacy_template_1 = require("./legacy-template");
/* eslint-disable max-len */
/**
 * Deploy legacy bootstrap stack
 *
 * @experimental
 */
async function bootstrapEnvironment(environment, sdkProvider, options = {}) {
    var _a, _b, _c;
    const params = (_a = options.parameters) !== null && _a !== void 0 ? _a : {};
    if ((_b = params.trustedAccounts) === null || _b === void 0 ? void 0 : _b.length) {
        throw new Error('--trust can only be passed for the new bootstrap experience.');
    }
    if ((_c = params.cloudFormationExecutionPolicies) === null || _c === void 0 ? void 0 : _c.length) {
        throw new Error('--cloudformation-execution-policies can only be passed for the new bootstrap experience.');
    }
    if (params.qualifier) {
        throw new Error('--qualifier can only be passed for the new bootstrap experience.');
    }
    return deploy_bootstrap_1.deployBootstrapStack(legacy_template_1.legacyBootstrapTemplate(params), {}, environment, sdkProvider, options);
}
exports.bootstrapEnvironment = bootstrapEnvironment;
/**
 * Deploy CI/CD-ready bootstrap stack from template
 *
 * @experimental
 */
async function bootstrapEnvironment2(environment, sdkProvider, options = {}) {
    var _a, _b, _c, _d, _e;
    const params = (_a = options.parameters) !== null && _a !== void 0 ? _a : {};
    if (((_b = params.trustedAccounts) === null || _b === void 0 ? void 0 : _b.length) && !((_c = params.cloudFormationExecutionPolicies) === null || _c === void 0 ? void 0 : _c.length)) {
        throw new Error('--cloudformation-execution-policies are required if --trust has been passed!');
    }
    const bootstrapTemplatePath = path.join(__dirname, 'bootstrap-template.yaml');
    const bootstrapTemplate = await serialize_1.loadStructuredFile(bootstrapTemplatePath);
    return deploy_bootstrap_1.deployBootstrapStack(bootstrapTemplate, {
        FileAssetsBucketName: params.bucketName,
        FileAssetsBucketKmsKeyId: params.kmsKeyId,
        TrustedAccounts: (_d = params.trustedAccounts) === null || _d === void 0 ? void 0 : _d.join(','),
        CloudFormationExecutionPolicies: (_e = params.cloudFormationExecutionPolicies) === null || _e === void 0 ? void 0 : _e.join(','),
        Qualifier: params.qualifier,
        PublicAccessBlockConfiguration: params.publicAccessBlockConfiguration || params.publicAccessBlockConfiguration === undefined ? 'true' : 'false',
    }, environment, sdkProvider, options);
}
exports.bootstrapEnvironment2 = bootstrapEnvironment2;
