// aws-cdk/test/context-providers/amis.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable import/order */
const aws = require("aws-sdk");
const AWS = require("aws-sdk-mock");
const ami_1 = require("../../lib/context-providers/ami");
const mock_sdk_1 = require("../util/mock-sdk");
// If the 'aws-sdk' package imported here and the 'aws-sdk' package imported by 'aws-sdk-mock' aren't
// the same physical package on disk (if version mismatches cause hoisting/deduping to not happen),
// the type check here takes too long and makes the TypeScript compiler fail.
// Suppress the type check using 'as any' to make this more robust.
AWS.setSDKInstance(aws);
afterEach(done => {
    AWS.restore();
    done();
});
const mockSDK = new mock_sdk_1.MockSdkProvider();
test('calls DescribeImages on the request', async () => {
    // GIVEN
    let request;
    AWS.mock('EC2', 'describeImages', (params, cb) => {
        request = params;
        return cb(null, { Images: [{ ImageId: 'ami-1234' }] });
    });
    // WHEN
    await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
        owners: ['some-owner'],
        filters: {
            'some-filter': ['filtered'],
        },
    });
    // THEN
    expect(request).toEqual({
        Owners: ['some-owner'],
        Filters: [
            {
                Name: 'some-filter',
                Values: ['filtered'],
            },
        ],
    });
});
test('returns the most recent AMI matching the criteria', async () => {
    // GIVEN
    AWS.mock('EC2', 'describeImages', (_, cb) => {
        return cb(null, {
            Images: [
                {
                    ImageId: 'ami-1234',
                    CreationDate: '2016-06-22T08:39:59.000Z',
                },
                {
                    ImageId: 'ami-5678',
                    CreationDate: '2019-06-22T08:39:59.000Z',
                },
            ],
        });
    });
    // WHEN
    const result = await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
        filters: {},
    });
    // THEN
    expect(result).toBe('ami-5678');
});
