// @aws-cdk/aws-apigateway/test/integ.restapi-import.lit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
/**
 * This file showcases how to split up a RestApi's Resources and Methods across nested stacks.
 *
 * The root stack 'RootStack' first defines a RestApi.
 * Two nested stacks BooksStack and PetsStack, create corresponding Resources '/books' and '/pets'.
 * They are then deployed to a 'prod' Stage via a third nested stack - DeployStack.
 *
 * To verify this worked, go to the APIGateway
 */
class RootStack extends core_1.Stack {
    constructor(scope) {
        super(scope, 'integ-restapi-import-RootStack');
        const restApi = new lib_1.RestApi(this, 'RestApi', {
            deploy: false,
        });
        restApi.root.addMethod('ANY');
        const petsStack = new PetsStack(this, {
            restApiId: restApi.restApiId,
            rootResourceId: restApi.restApiRootResourceId,
        });
        const booksStack = new BooksStack(this, {
            restApiId: restApi.restApiId,
            rootResourceId: restApi.restApiRootResourceId,
        });
        new DeployStack(this, {
            restApiId: restApi.restApiId,
            methods: [...petsStack.methods, ...booksStack.methods],
        });
        new core_1.CfnOutput(this, 'PetsURL', {
            value: `https://${restApi.restApiId}.execute-api.${this.region}.amazonaws.com/prod/pets`,
        });
        new core_1.CfnOutput(this, 'BooksURL', {
            value: `https://${restApi.restApiId}.execute-api.${this.region}.amazonaws.com/prod/books`,
        });
    }
}
class PetsStack extends core_1.NestedStack {
    constructor(scope, props) {
        super(scope, 'integ-restapi-import-PetsStack', props);
        this.methods = [];
        const api = lib_1.RestApi.fromRestApiAttributes(this, 'RestApi', {
            restApiId: props.restApiId,
            rootResourceId: props.rootResourceId,
        });
        const method = api.root.addResource('pets').addMethod('GET', new lib_1.MockIntegration({
            integrationResponses: [{
                    statusCode: '200',
                }],
            passthroughBehavior: lib_1.PassthroughBehavior.NEVER,
            requestTemplates: {
                'application/json': '{ "statusCode": 200 }',
            },
        }), {
            methodResponses: [{ statusCode: '200' }],
        });
        this.methods.push(method);
    }
}
class BooksStack extends core_1.NestedStack {
    constructor(scope, props) {
        super(scope, 'integ-restapi-import-BooksStack', props);
        this.methods = [];
        const api = lib_1.RestApi.fromRestApiAttributes(this, 'RestApi', {
            restApiId: props.restApiId,
            rootResourceId: props.rootResourceId,
        });
        const method = api.root.addResource('books').addMethod('GET', new lib_1.MockIntegration({
            integrationResponses: [{
                    statusCode: '200',
                }],
            passthroughBehavior: lib_1.PassthroughBehavior.NEVER,
            requestTemplates: {
                'application/json': '{ "statusCode": 200 }',
            },
        }), {
            methodResponses: [{ statusCode: '200' }],
        });
        this.methods.push(method);
    }
}
class DeployStack extends core_1.NestedStack {
    constructor(scope, props) {
        var _a;
        super(scope, 'integ-restapi-import-DeployStack', props);
        const deployment = new lib_1.Deployment(this, 'Deployment', {
            api: lib_1.RestApi.fromRestApiId(this, 'RestApi', props.restApiId),
        });
        ((_a = props.methods) !== null && _a !== void 0 ? _a : []).forEach((method) => deployment.node.addDependency(method));
        new lib_1.Stage(this, 'Stage', { deployment });
    }
}
new RootStack(new core_1.App());
