// @aws-cdk/core/test/custom-resource-provider/test.custom-resource-provider.js
"use strict";
const fs = require("fs");
const path = require("path");
const lib_1 = require("../../lib");
const util_1 = require("../util");
const TEST_HANDLER = `${__dirname}/mock-provider`;
module.exports = {
    'minimal configuration'(test) {
        var _a;
        // GIVEN
        const stack = new lib_1.Stack();
        // WHEN
        lib_1.CustomResourceProvider.getOrCreate(stack, 'Custom:MyResourceType', {
            codeDirectory: TEST_HANDLER,
            runtime: lib_1.CustomResourceProviderRuntime.NODEJS_12,
        });
        // THEN
        test.ok(fs.existsSync(path.join(TEST_HANDLER, '__entrypoint__.js')), 'expecting entrypoint to be copied to the handler directory');
        const cfn = util_1.toCloudFormation(stack);
        // The asset hash constantly changes, so in order to not have to chase it, just look
        // it up from the output.
        const staging = (_a = stack.node.tryFindChild('Custom:MyResourceTypeCustomResourceProvider')) === null || _a === void 0 ? void 0 : _a.node.tryFindChild('Staging');
        const assetHash = staging.sourceHash;
        const paramNames = Object.keys(cfn.Parameters);
        const bucketParam = paramNames[0];
        const keyParam = paramNames[1];
        const hashParam = paramNames[2];
        test.deepEqual(cfn, {
            Resources: {
                CustomMyResourceTypeCustomResourceProviderRoleBD5E655F: {
                    Type: 'AWS::IAM::Role',
                    Properties: {
                        AssumeRolePolicyDocument: {
                            Version: '2012-10-17',
                            Statement: [
                                {
                                    Action: 'sts:AssumeRole',
                                    Effect: 'Allow',
                                    Principal: {
                                        Service: 'lambda.amazonaws.com',
                                    },
                                },
                            ],
                        },
                        ManagedPolicyArns: [
                            {
                                'Fn::Sub': 'arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
                            },
                        ],
                    },
                },
                CustomMyResourceTypeCustomResourceProviderHandler29FBDD2A: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Code: {
                            S3Bucket: { Ref: bucketParam },
                            S3Key: {
                                'Fn::Join': [
                                    '',
                                    [
                                        {
                                            'Fn::Select': [
                                                0,
                                                {
                                                    'Fn::Split': [
                                                        '||',
                                                        { Ref: keyParam },
                                                    ],
                                                },
                                            ],
                                        },
                                        {
                                            'Fn::Select': [
                                                1,
                                                {
                                                    'Fn::Split': [
                                                        '||',
                                                        { Ref: keyParam },
                                                    ],
                                                },
                                            ],
                                        },
                                    ],
                                ],
                            },
                        },
                        Timeout: 900,
                        MemorySize: 128,
                        Handler: '__entrypoint__.handler',
                        Role: {
                            'Fn::GetAtt': [
                                'CustomMyResourceTypeCustomResourceProviderRoleBD5E655F',
                                'Arn',
                            ],
                        },
                        Runtime: 'nodejs12.x',
                    },
                    DependsOn: [
                        'CustomMyResourceTypeCustomResourceProviderRoleBD5E655F',
                    ],
                },
            },
            Parameters: {
                [bucketParam]: {
                    Type: 'String',
                    Description: `S3 bucket for asset "${assetHash}"`,
                },
                [keyParam]: {
                    Type: 'String',
                    Description: `S3 key for asset version "${assetHash}"`,
                },
                [hashParam]: {
                    Type: 'String',
                    Description: `Artifact hash for asset "${assetHash}"`,
                },
            },
        });
        test.done();
    },
    'policyStatements can be used to add statements to the inline policy'(test) {
        // GIVEN
        const stack = new lib_1.Stack();
        // WHEN
        lib_1.CustomResourceProvider.getOrCreate(stack, 'Custom:MyResourceType', {
            codeDirectory: TEST_HANDLER,
            runtime: lib_1.CustomResourceProviderRuntime.NODEJS_12,
            policyStatements: [
                { statement1: 123 },
                { statement2: { foo: 111 } },
            ],
        });
        // THEN
        const template = util_1.toCloudFormation(stack);
        const role = template.Resources.CustomMyResourceTypeCustomResourceProviderRoleBD5E655F;
        test.deepEqual(role.Properties.Policies, [{
                PolicyName: 'Inline',
                PolicyDocument: {
                    Version: '2012-10-17',
                    Statement: [{ statement1: 123 }, { statement2: { foo: 111 } }],
                },
            }]);
        test.done();
    },
    'memorySize and timeout'(test) {
        // GIVEN
        const stack = new lib_1.Stack();
        // WHEN
        lib_1.CustomResourceProvider.getOrCreate(stack, 'Custom:MyResourceType', {
            codeDirectory: TEST_HANDLER,
            runtime: lib_1.CustomResourceProviderRuntime.NODEJS_12,
            memorySize: lib_1.Size.gibibytes(2),
            timeout: lib_1.Duration.minutes(5),
        });
        // THEN
        const template = util_1.toCloudFormation(stack);
        const lambda = template.Resources.CustomMyResourceTypeCustomResourceProviderHandler29FBDD2A;
        test.deepEqual(lambda.Properties.MemorySize, 2048);
        test.deepEqual(lambda.Properties.Timeout, 300);
        test.done();
    },
};
