// @aws-cdk/aws-cognito/test/user-pool-idps/amazon.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../../lib");
describe('UserPoolIdentityProvider', () => {
    describe('amazon', () => {
        test('defaults', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            new lib_1.UserPoolIdentityProviderAmazon(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'amzn-client-id',
                clientSecret: 'amzn-client-secret',
            });
            expect(stack).toHaveResource('AWS::Cognito::UserPoolIdentityProvider', {
                ProviderName: 'LoginWithAmazon',
                ProviderType: 'LoginWithAmazon',
                ProviderDetails: {
                    client_id: 'amzn-client-id',
                    client_secret: 'amzn-client-secret',
                    authorize_scopes: 'profile',
                },
            });
        });
        test('scopes', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            new lib_1.UserPoolIdentityProviderAmazon(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'amzn-client-id',
                clientSecret: 'amzn-client-secret',
                scopes: ['scope1', 'scope2'],
            });
            expect(stack).toHaveResource('AWS::Cognito::UserPoolIdentityProvider', {
                ProviderName: 'LoginWithAmazon',
                ProviderType: 'LoginWithAmazon',
                ProviderDetails: {
                    client_id: 'amzn-client-id',
                    client_secret: 'amzn-client-secret',
                    authorize_scopes: 'scope1 scope2',
                },
            });
        });
        test('registered with user pool', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            const provider = new lib_1.UserPoolIdentityProviderAmazon(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'amzn-client-id',
                clientSecret: 'amzn-client-secret',
            });
            // THEN
            expect(pool.identityProviders).toContain(provider);
        });
        test('attribute mapping', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            new lib_1.UserPoolIdentityProviderAmazon(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'amazn-client-id',
                clientSecret: 'amzn-client-secret',
                attributeMapping: {
                    givenName: lib_1.ProviderAttribute.AMAZON_NAME,
                    address: lib_1.ProviderAttribute.other('amzn-address'),
                    custom: {
                        customAttr1: lib_1.ProviderAttribute.AMAZON_EMAIL,
                        customAttr2: lib_1.ProviderAttribute.other('amzn-custom-attr'),
                    },
                },
            });
            // THEN
            expect(stack).toHaveResource('AWS::Cognito::UserPoolIdentityProvider', {
                AttributeMapping: {
                    given_name: 'name',
                    address: 'amzn-address',
                    customAttr1: 'email',
                    customAttr2: 'amzn-custom-attr',
                },
            });
        });
    });
});
