// @aws-cdk/aws-cloudfront/test/private/cache-behavior.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../../lib");
const cache_behavior_1 = require("../../lib/private/cache-behavior");
let app;
beforeEach(() => {
    app = new core_1.App();
    new core_1.Stack(app, 'Stack', {
        env: { account: '1234', region: 'testregion' },
    });
});
test('renders the minimum template with an origin and path specified', () => {
    const behavior = new cache_behavior_1.CacheBehavior('origin_id', {
        pathPattern: '*',
    });
    expect(behavior._renderBehavior()).toEqual({
        targetOriginId: 'origin_id',
        pathPattern: '*',
        forwardedValues: { queryString: false },
        viewerProtocolPolicy: 'allow-all',
    });
});
test('renders with all properties specified', () => {
    const behavior = new cache_behavior_1.CacheBehavior('origin_id', {
        pathPattern: '*',
        allowedMethods: lib_1.AllowedMethods.ALLOW_ALL,
        cachedMethods: lib_1.CachedMethods.CACHE_GET_HEAD_OPTIONS,
        compress: true,
        forwardQueryString: true,
        forwardQueryStringCacheKeys: ['user_id', 'auth'],
        smoothStreaming: true,
        viewerProtocolPolicy: lib_1.ViewerProtocolPolicy.HTTPS_ONLY,
    });
    expect(behavior._renderBehavior()).toEqual({
        targetOriginId: 'origin_id',
        pathPattern: '*',
        allowedMethods: ['GET', 'HEAD', 'OPTIONS', 'PUT', 'PATCH', 'POST', 'DELETE'],
        cachedMethods: ['GET', 'HEAD', 'OPTIONS'],
        compress: true,
        forwardedValues: {
            queryString: true,
            queryStringCacheKeys: ['user_id', 'auth'],
        },
        smoothStreaming: true,
        viewerProtocolPolicy: 'https-only',
    });
});
