// aws-cdk/test/integ/cli/cli.integtest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const os = require("os");
const path = require("path");
const aws_helpers_1 = require("./aws-helpers");
const cdk_helpers_1 = require("./cdk-helpers");
const test_helpers_1 = require("./test-helpers");
jest.setTimeout(600 * 1000);
beforeAll(async () => {
    await cdk_helpers_1.prepareAppFixture();
});
beforeEach(async () => {
    await cdk_helpers_1.cleanup();
});
afterEach(async () => {
    await cdk_helpers_1.cleanup();
});
test_helpers_1.integTest('VPC Lookup', async () => {
    cdk_helpers_1.log('Making sure we are clean before starting.');
    await cdk_helpers_1.cdkDestroy('define-vpc', { modEnv: { ENABLE_VPC_TESTING: 'DEFINE' } });
    cdk_helpers_1.log('Setting up: creating a VPC with known tags');
    await cdk_helpers_1.cdkDeploy('define-vpc', { modEnv: { ENABLE_VPC_TESTING: 'DEFINE' } });
    cdk_helpers_1.log('Setup complete!');
    cdk_helpers_1.log('Verifying we can now import that VPC');
    await cdk_helpers_1.cdkDeploy('import-vpc', { modEnv: { ENABLE_VPC_TESTING: 'IMPORT' } });
});
test_helpers_1.integTest('Two ways of shoing the version', async () => {
    const version1 = await cdk_helpers_1.cdk(['version']);
    const version2 = await cdk_helpers_1.cdk(['--version']);
    expect(version1).toEqual(version2);
});
test_helpers_1.integTest('Termination protection', async () => {
    const stackName = 'termination-protection';
    await cdk_helpers_1.cdkDeploy(stackName);
    // Try a destroy that should fail
    await expect(cdk_helpers_1.cdkDestroy(stackName)).rejects.toThrow('exited with error');
    // Can update termination protection even though the change set doesn't contain changes
    await cdk_helpers_1.cdkDeploy(stackName, { modEnv: { TERMINATION_PROTECTION: 'FALSE' } });
    await cdk_helpers_1.cdkDestroy(stackName);
});
test_helpers_1.integTest('cdk synth', async () => {
    await expect(cdk_helpers_1.cdk(['synth', cdk_helpers_1.fullStackName('test-1')])).resolves.toEqual(`Resources:
  topic69831491:
    Type: AWS::SNS::Topic
    Metadata:
      aws:cdk:path: ${cdk_helpers_1.STACK_NAME_PREFIX}-test-1/topic/Resource`);
    await expect(cdk_helpers_1.cdk(['synth', cdk_helpers_1.fullStackName('test-2')])).resolves.toEqual(`Resources:
  topic152D84A37:
    Type: AWS::SNS::Topic
    Metadata:
      aws:cdk:path: ${cdk_helpers_1.STACK_NAME_PREFIX}-test-2/topic1/Resource
  topic2A4FB547F:
    Type: AWS::SNS::Topic
    Metadata:
      aws:cdk:path: ${cdk_helpers_1.STACK_NAME_PREFIX}-test-2/topic2/Resource`);
});
test_helpers_1.integTest('ssm parameter provider error', async () => {
    await expect(cdk_helpers_1.cdk(['synth',
        cdk_helpers_1.fullStackName('missing-ssm-parameter'),
        '-c', 'test:ssm-parameter-name=/does/not/exist',
    ], {
        allowErrExit: true,
    })).resolves.toContain('SSM parameter not available in account');
});
test_helpers_1.integTest('automatic ordering', async () => {
    // Deploy the consuming stack which will include the producing stack
    await cdk_helpers_1.cdkDeploy('order-consuming');
    // Destroy the providing stack which will include the consuming stack
    await cdk_helpers_1.cdkDestroy('order-providing');
});
test_helpers_1.integTest('context setting', async () => {
    await fs_1.promises.writeFile(path.join(cdk_helpers_1.INTEG_TEST_DIR, 'cdk.context.json'), JSON.stringify({
        contextkey: 'this is the context value',
    }));
    try {
        await expect(cdk_helpers_1.cdk(['context'])).resolves.toContain('this is the context value');
        // Test that deleting the contextkey works
        await cdk_helpers_1.cdk(['context', '--reset', 'contextkey']);
        await expect(cdk_helpers_1.cdk(['context'])).resolves.not.toContain('this is the context value');
        // Test that forced delete of the context key does not throw
        await cdk_helpers_1.cdk(['context', '-f', '--reset', 'contextkey']);
    }
    finally {
        await fs_1.promises.unlink(path.join(cdk_helpers_1.INTEG_TEST_DIR, 'cdk.context.json'));
    }
});
test_helpers_1.integTest('deploy', async () => {
    var _a;
    const stackArn = await cdk_helpers_1.cdkDeploy('test-2', { captureStderr: false });
    // verify the number of resources in the stack
    const response = await aws_helpers_1.cloudFormation('describeStackResources', {
        StackName: stackArn,
    });
    expect((_a = response.StackResources) === null || _a === void 0 ? void 0 : _a.length).toEqual(2);
});
test_helpers_1.integTest('deploy all', async () => {
    const arns = await cdk_helpers_1.cdkDeploy('test-*', { captureStderr: false });
    // verify that we only deployed a single stack (there's a single ARN in the output)
    expect(arns.split('\n').length).toEqual(2);
});
test_helpers_1.integTest('nested stack with parameters', async () => {
    var _a;
    // STACK_NAME_PREFIX is used in MyTopicParam to allow multiple instances
    // of this test to run in parallel, othewise they will attempt to create the same SNS topic.
    const stackArn = await cdk_helpers_1.cdkDeploy('with-nested-stack-using-parameters', {
        options: ['--parameters', 'MyTopicParam=${STACK_NAME_PREFIX}ThereIsNoSpoon'],
        captureStderr: false,
    });
    // verify that we only deployed a single stack (there's a single ARN in the output)
    expect(stackArn.split('\n').length).toEqual(1);
    // verify the number of resources in the stack
    const response = await aws_helpers_1.cloudFormation('describeStackResources', {
        StackName: stackArn,
    });
    expect((_a = response.StackResources) === null || _a === void 0 ? void 0 : _a.length).toEqual(1);
});
test_helpers_1.integTest('deploy without execute', async () => {
    var _a;
    const stackArn = await cdk_helpers_1.cdkDeploy('test-2', {
        options: ['--no-execute'],
        captureStderr: false,
    });
    // verify that we only deployed a single stack (there's a single ARN in the output)
    expect(stackArn.split('\n').length).toEqual(1);
    const response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: stackArn,
    });
    expect((_a = response.Stacks) === null || _a === void 0 ? void 0 : _a[0].StackStatus).toEqual('REVIEW_IN_PROGRESS');
});
test_helpers_1.integTest('security related changes without a CLI are expected to fail', async () => {
    // redirect /dev/null to stdin, which means there will not be tty attached
    // since this stack includes security-related changes, the deployment should
    // immediately fail because we can't confirm the changes
    const stackName = 'iam-test';
    await expect(cdk_helpers_1.cdkDeploy(stackName, {
        options: ['<', '/dev/null'],
        neverRequireApproval: false,
    })).rejects.toThrow('exited with error');
    // Ensure stack was not deployed
    await expect(aws_helpers_1.cloudFormation('describeStacks', {
        StackName: cdk_helpers_1.fullStackName(stackName),
    })).rejects.toThrow('does not exist');
});
test_helpers_1.integTest('deploy wildcard with outputs', async () => {
    const outputsFile = path.join(cdk_helpers_1.INTEG_TEST_DIR, 'outputs', 'outputs.json');
    await fs_1.promises.mkdir(path.dirname(outputsFile), { recursive: true });
    await cdk_helpers_1.cdkDeploy(['outputs-test-*'], {
        options: ['--outputs-file', outputsFile],
    });
    const outputs = JSON.parse((await fs_1.promises.readFile(outputsFile, { encoding: 'utf-8' })).toString());
    expect(outputs).toEqual({
        [`${cdk_helpers_1.STACK_NAME_PREFIX}-outputs-test-1`]: {
            TopicName: `${cdk_helpers_1.STACK_NAME_PREFIX}-outputs-test-1MyTopic`,
        },
        [`${cdk_helpers_1.STACK_NAME_PREFIX}-outputs-test-2`]: {
            TopicName: `${cdk_helpers_1.STACK_NAME_PREFIX}-outputs-test-2MyOtherTopic`,
        },
    });
});
test_helpers_1.integTest('deploy with parameters', async () => {
    var _a;
    const stackArn = await cdk_helpers_1.cdkDeploy('param-test-1', {
        options: [
            '--parameters', `TopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}bazinga`,
        ],
        captureStderr: false,
    });
    const response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: stackArn,
    });
    expect((_a = response.Stacks) === null || _a === void 0 ? void 0 : _a[0].Parameters).toEqual([
        {
            ParameterKey: 'TopicNameParam',
            ParameterValue: `${cdk_helpers_1.STACK_NAME_PREFIX}bazinga`,
        },
    ]);
});
test_helpers_1.integTest('update to stack in ROLLBACK_COMPLETE state will delete stack and create a new one', async () => {
    var _a, _b, _c, _d;
    // GIVEN
    await expect(cdk_helpers_1.cdkDeploy('param-test-1', {
        options: [
            '--parameters', `TopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}@aww`,
        ],
        captureStderr: false,
    })).rejects.toThrow('exited with error');
    const response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: cdk_helpers_1.fullStackName('param-test-1'),
    });
    const stackArn = (_a = response.Stacks) === null || _a === void 0 ? void 0 : _a[0].StackId;
    expect((_b = response.Stacks) === null || _b === void 0 ? void 0 : _b[0].StackStatus).toEqual('ROLLBACK_COMPLETE');
    // WHEN
    const newStackArn = await cdk_helpers_1.cdkDeploy('param-test-1', {
        options: [
            '--parameters', `TopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}allgood`,
        ],
        captureStderr: false,
    });
    const newStackResponse = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: newStackArn,
    });
    // THEN
    expect(stackArn).not.toEqual(newStackArn); // new stack was created
    expect((_c = newStackResponse.Stacks) === null || _c === void 0 ? void 0 : _c[0].StackStatus).toEqual('CREATE_COMPLETE');
    expect((_d = newStackResponse.Stacks) === null || _d === void 0 ? void 0 : _d[0].Parameters).toEqual([
        {
            ParameterKey: 'TopicNameParam',
            ParameterValue: `${cdk_helpers_1.STACK_NAME_PREFIX}allgood`,
        },
    ]);
});
test_helpers_1.integTest('stack in UPDATE_ROLLBACK_COMPLETE state can be updated', async () => {
    var _a, _b, _c, _d;
    // GIVEN
    const stackArn = await cdk_helpers_1.cdkDeploy('param-test-1', {
        options: [
            '--parameters', `TopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}nice`,
        ],
        captureStderr: false,
    });
    let response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: stackArn,
    });
    expect((_a = response.Stacks) === null || _a === void 0 ? void 0 : _a[0].StackStatus).toEqual('CREATE_COMPLETE');
    // bad parameter name with @ will put stack into UPDATE_ROLLBACK_COMPLETE
    await expect(cdk_helpers_1.cdkDeploy('param-test-1', {
        options: [
            '--parameters', `TopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}@aww`,
        ],
        captureStderr: false,
    })).rejects.toThrow('exited with error');
    ;
    response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: stackArn,
    });
    expect((_b = response.Stacks) === null || _b === void 0 ? void 0 : _b[0].StackStatus).toEqual('UPDATE_ROLLBACK_COMPLETE');
    // WHEN
    await cdk_helpers_1.cdkDeploy('param-test-1', {
        options: [
            '--parameters', `TopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}allgood`,
        ],
        captureStderr: false,
    });
    response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: stackArn,
    });
    // THEN
    expect((_c = response.Stacks) === null || _c === void 0 ? void 0 : _c[0].StackStatus).toEqual('UPDATE_COMPLETE');
    expect((_d = response.Stacks) === null || _d === void 0 ? void 0 : _d[0].Parameters).toEqual([
        {
            ParameterKey: 'TopicNameParam',
            ParameterValue: `${cdk_helpers_1.STACK_NAME_PREFIX}allgood`,
        },
    ]);
});
test_helpers_1.integTest('deploy with wildcard and parameters', async () => {
    await cdk_helpers_1.cdkDeploy('param-test-*', {
        options: [
            '--parameters', `${cdk_helpers_1.STACK_NAME_PREFIX}-param-test-1:TopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}bazinga`,
            '--parameters', `${cdk_helpers_1.STACK_NAME_PREFIX}-param-test-2:OtherTopicNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}ThatsMySpot`,
            '--parameters', `${cdk_helpers_1.STACK_NAME_PREFIX}-param-test-3:DisplayNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}HeyThere`,
            '--parameters', `${cdk_helpers_1.STACK_NAME_PREFIX}-param-test-3:OtherDisplayNameParam=${cdk_helpers_1.STACK_NAME_PREFIX}AnotherOne`,
        ],
    });
});
test_helpers_1.integTest('deploy with parameters multi', async () => {
    var _a;
    const paramVal1 = `${cdk_helpers_1.STACK_NAME_PREFIX}bazinga`;
    const paramVal2 = `${cdk_helpers_1.STACK_NAME_PREFIX}=jagshemash`;
    const stackArn = await cdk_helpers_1.cdkDeploy('param-test-3', {
        options: [
            '--parameters', `DisplayNameParam=${paramVal1}`,
            '--parameters', `OtherDisplayNameParam=${paramVal2}`,
        ],
        captureStderr: false,
    });
    const response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: stackArn,
    });
    expect((_a = response.Stacks) === null || _a === void 0 ? void 0 : _a[0].Parameters).toEqual([
        {
            ParameterKey: 'DisplayNameParam',
            ParameterValue: paramVal1,
        },
        {
            ParameterKey: 'OtherDisplayNameParam',
            ParameterValue: paramVal2,
        },
    ]);
});
test_helpers_1.integTest('deploy with notification ARN', async () => {
    var _a;
    const topicName = `${cdk_helpers_1.STACK_NAME_PREFIX}-test-topic`;
    const response = await aws_helpers_1.sns('createTopic', { Name: topicName });
    const topicArn = response.TopicArn;
    try {
        await cdk_helpers_1.cdkDeploy('test-2', {
            options: ['--notification-arns', topicArn],
        });
        // verify that the stack we deployed has our notification ARN
        const describeResponse = await aws_helpers_1.cloudFormation('describeStacks', {
            StackName: cdk_helpers_1.fullStackName('test-2'),
        });
        expect((_a = describeResponse.Stacks) === null || _a === void 0 ? void 0 : _a[0].NotificationARNs).toEqual([topicArn]);
    }
    finally {
        await aws_helpers_1.sns('deleteTopic', {
            TopicArn: topicArn,
        });
    }
});
test_helpers_1.integTest('deploy with role', async () => {
    const roleName = `${cdk_helpers_1.STACK_NAME_PREFIX}-test-role`;
    await deleteRole();
    const createResponse = await aws_helpers_1.iam('createRole', {
        RoleName: roleName,
        AssumeRolePolicyDocument: JSON.stringify({
            Version: '2012-10-17',
            Statement: [{
                    Action: 'sts:AssumeRole',
                    Principal: { Service: 'cloudformation.amazonaws.com' },
                    Effect: 'Allow',
                }, {
                    Action: 'sts:AssumeRole',
                    Principal: { AWS: (await aws_helpers_1.sts('getCallerIdentity', {})).Arn },
                    Effect: 'Allow',
                }],
        }),
    });
    const roleArn = createResponse.Role.Arn;
    try {
        await aws_helpers_1.iam('putRolePolicy', {
            RoleName: roleName,
            PolicyName: 'DefaultPolicy',
            PolicyDocument: JSON.stringify({
                Version: '2012-10-17',
                Statement: [{
                        Action: '*',
                        Resource: '*',
                        Effect: 'Allow',
                    }],
            }),
        });
        await aws_helpers_1.retry('Trying to assume fresh role', aws_helpers_1.retry.forSeconds(300), async () => {
            await aws_helpers_1.sts('assumeRole', {
                RoleArn: roleArn,
                RoleSessionName: 'testing',
            });
        });
        // In principle, the role has replicated from 'us-east-1' to wherever we're testing.
        // Give it a little more sleep to make sure CloudFormation is not hitting a box
        // that doesn't have it yet.
        await aws_helpers_1.sleep(5000);
        await cdk_helpers_1.cdkDeploy('test-2', {
            options: ['--role-arn', roleArn],
        });
        // Immediately delete the stack again before we delete the role.
        //
        // Since roles are sticky, if we delete the role before the stack, subsequent DeleteStack
        // operations will fail when CloudFormation tries to assume the role that's already gone.
        await cdk_helpers_1.cdkDestroy('test-2');
    }
    finally {
        await deleteRole();
    }
    async function deleteRole() {
        try {
            for (const policyName of (await aws_helpers_1.iam('listRolePolicies', { RoleName: roleName })).PolicyNames) {
                await aws_helpers_1.iam('deleteRolePolicy', {
                    RoleName: roleName,
                    PolicyName: policyName,
                });
            }
            await aws_helpers_1.iam('deleteRole', { RoleName: roleName });
        }
        catch (e) {
            if (e.message.indexOf('cannot be found') > -1) {
                return;
            }
            throw e;
        }
    }
});
test_helpers_1.integTest('cdk diff', async () => {
    const diff1 = await cdk_helpers_1.cdk(['diff', cdk_helpers_1.fullStackName('test-1')]);
    expect(diff1).toContain('AWS::SNS::Topic');
    const diff2 = await cdk_helpers_1.cdk(['diff', cdk_helpers_1.fullStackName('test-2')]);
    expect(diff2).toContain('AWS::SNS::Topic');
    // We can make it fail by passing --fail
    await expect(cdk_helpers_1.cdk(['diff', '--fail', cdk_helpers_1.fullStackName('test-1')]))
        .rejects.toThrow('exited with error');
});
test_helpers_1.integTest('cdk diff --fail on multiple stacks exits with error if any of the stacks contains a diff', async () => {
    // GIVEN
    const diff1 = await cdk_helpers_1.cdk(['diff', cdk_helpers_1.fullStackName('test-1')]);
    expect(diff1).toContain('AWS::SNS::Topic');
    await cdk_helpers_1.cdkDeploy('test-2');
    const diff2 = await cdk_helpers_1.cdk(['diff', cdk_helpers_1.fullStackName('test-2')]);
    expect(diff2).toContain('There were no differences');
    // WHEN / THEN
    await expect(cdk_helpers_1.cdk(['diff', '--fail', cdk_helpers_1.fullStackName('test-1'), cdk_helpers_1.fullStackName('test-2')])).rejects.toThrow('exited with error');
});
test_helpers_1.integTest('cdk diff --fail with multiple stack exits with if any of the stacks contains a diff', async () => {
    // GIVEN
    await cdk_helpers_1.cdkDeploy('test-1');
    const diff1 = await cdk_helpers_1.cdk(['diff', cdk_helpers_1.fullStackName('test-1')]);
    expect(diff1).toContain('There were no differences');
    const diff2 = await cdk_helpers_1.cdk(['diff', cdk_helpers_1.fullStackName('test-2')]);
    expect(diff2).toContain('AWS::SNS::Topic');
    // WHEN / THEN
    await expect(cdk_helpers_1.cdk(['diff', '--fail', cdk_helpers_1.fullStackName('test-1'), cdk_helpers_1.fullStackName('test-2')])).rejects.toThrow('exited with error');
});
test_helpers_1.integTest('deploy stack with docker asset', async () => {
    await cdk_helpers_1.cdkDeploy('docker');
});
test_helpers_1.integTest('deploy and test stack with lambda asset', async () => {
    var _a, _b;
    const stackArn = await cdk_helpers_1.cdkDeploy('lambda', { captureStderr: false });
    const response = await aws_helpers_1.cloudFormation('describeStacks', {
        StackName: stackArn,
    });
    const lambdaArn = (_b = (_a = response.Stacks) === null || _a === void 0 ? void 0 : _a[0].Outputs) === null || _b === void 0 ? void 0 : _b[0].OutputValue;
    if (lambdaArn === undefined) {
        throw new Error('Stack did not have expected Lambda ARN output');
    }
    const output = await aws_helpers_1.lambda('invoke', {
        FunctionName: lambdaArn,
    });
    expect(JSON.stringify(output.Payload)).toContain('dear asset');
});
test_helpers_1.integTest('cdk ls', async () => {
    const listing = await cdk_helpers_1.cdk(['ls'], { captureStderr: false });
    const expectedStacks = [
        'conditional-resource',
        'docker',
        'docker-with-custom-file',
        'failed',
        'iam-test',
        'lambda',
        'missing-ssm-parameter',
        'order-providing',
        'outputs-test-1',
        'outputs-test-2',
        'param-test-1',
        'param-test-2',
        'param-test-3',
        'termination-protection',
        'test-1',
        'test-2',
        'with-nested-stack',
        'with-nested-stack-using-parameters',
        'order-consuming',
    ];
    for (const stack of expectedStacks) {
        expect(listing).toContain(cdk_helpers_1.fullStackName(stack));
    }
});
test_helpers_1.integTest('deploy stack without resource', async () => {
    // Deploy the stack without resources
    await cdk_helpers_1.cdkDeploy('conditional-resource', { modEnv: { NO_RESOURCE: 'TRUE' } });
    // This should have succeeded but not deployed the stack.
    await expect(aws_helpers_1.cloudFormation('describeStacks', { StackName: cdk_helpers_1.fullStackName('conditional-resource') }))
        .rejects.toThrow('conditional-resource does not exist');
    // Deploy the stack with resources
    await cdk_helpers_1.cdkDeploy('conditional-resource');
    // Then again WITHOUT resources (this should destroy the stack)
    await cdk_helpers_1.cdkDeploy('conditional-resource', { modEnv: { NO_RESOURCE: 'TRUE' } });
    await expect(aws_helpers_1.cloudFormation('describeStacks', { StackName: cdk_helpers_1.fullStackName('conditional-resource') }))
        .rejects.toThrow('conditional-resource does not exist');
});
test_helpers_1.integTest('IAM diff', async () => {
    const output = await cdk_helpers_1.cdk(['diff', cdk_helpers_1.fullStackName('iam-test')]);
    // Roughly check for a table like this:
    //
    // ┌───┬─────────────────┬────────┬────────────────┬────────────────────────────-──┬───────────┐
    // │   │ Resource        │ Effect │ Action         │ Principal                     │ Condition │
    // ├───┼─────────────────┼────────┼────────────────┼───────────────────────────────┼───────────┤
    // │ + │ ${SomeRole.Arn} │ Allow  │ sts:AssumeRole │ Service:ec2.amazonaws.com     │           │
    // └───┴─────────────────┴────────┴────────────────┴───────────────────────────────┴───────────┘
    expect(output).toContain('${SomeRole.Arn}');
    expect(output).toContain('sts:AssumeRole');
    expect(output).toContain('ec2.amazonaws.com');
});
test_helpers_1.integTest('fast deploy', async () => {
    // we are using a stack with a nested stack because CFN will always attempt to
    // update a nested stack, which will allow us to verify that updates are actually
    // skipped unless --force is specified.
    const stackArn = await cdk_helpers_1.cdkDeploy('with-nested-stack', { captureStderr: false });
    const changeSet1 = await getLatestChangeSet();
    // Deploy the same stack again, there should be no new change set created
    await cdk_helpers_1.cdkDeploy('with-nested-stack');
    const changeSet2 = await getLatestChangeSet();
    expect(changeSet2.ChangeSetId).toEqual(changeSet1.ChangeSetId);
    // Deploy the stack again with --force, now we should create a changeset
    await cdk_helpers_1.cdkDeploy('with-nested-stack', { options: ['--force'] });
    const changeSet3 = await getLatestChangeSet();
    expect(changeSet3.ChangeSetId).not.toEqual(changeSet2.ChangeSetId);
    // Deploy the stack again with tags, expected to create a new changeset
    // even though the resources didn't change.
    await cdk_helpers_1.cdkDeploy('with-nested-stack', { options: ['--tags', 'key=value'] });
    const changeSet4 = await getLatestChangeSet();
    expect(changeSet4.ChangeSetId).not.toEqual(changeSet3.ChangeSetId);
    async function getLatestChangeSet() {
        var _a, _b, _c;
        const response = await aws_helpers_1.cloudFormation('describeStacks', { StackName: stackArn });
        if (!((_a = response.Stacks) === null || _a === void 0 ? void 0 : _a[0])) {
            throw new Error('Did not get a ChangeSet at all');
        }
        cdk_helpers_1.log(`Found Change Set ${(_b = response.Stacks) === null || _b === void 0 ? void 0 : _b[0].ChangeSetId}`);
        return (_c = response.Stacks) === null || _c === void 0 ? void 0 : _c[0];
    }
});
test_helpers_1.integTest('failed deploy does not hang', async () => {
    // this will hang if we introduce https://github.com/aws/aws-cdk/issues/6403 again.
    await expect(cdk_helpers_1.cdkDeploy('failed')).rejects.toThrow('exited with error');
});
test_helpers_1.integTest('can still load old assemblies', async () => {
    const cxAsmDir = path.join(os.tmpdir(), 'cdk-integ-cx');
    const testAssembliesDirectory = path.join(__dirname, 'cloud-assemblies');
    for (const asmdir of await listChildDirs(testAssembliesDirectory)) {
        cdk_helpers_1.log(`ASSEMBLY ${asmdir}`);
        await cdk_helpers_1.cloneDirectory(asmdir, cxAsmDir);
        // Some files in the asm directory that have a .js extension are
        // actually treated as templates. Evaluate them using NodeJS.
        const templates = await listChildren(cxAsmDir, fullPath => Promise.resolve(fullPath.endsWith('.js')));
        for (const template of templates) {
            const targetName = template.replace(/.js$/, '');
            await cdk_helpers_1.shell([process.execPath, template, '>', targetName], {
                cwd: cxAsmDir,
                modEnv: {
                    TEST_ACCOUNT: (await aws_helpers_1.testEnv()).account,
                    TEST_REGION: (await aws_helpers_1.testEnv()).region,
                },
            });
        }
        // Use this directory as a Cloud Assembly
        const output = await cdk_helpers_1.cdk([
            '--app', cxAsmDir,
            '-v',
            'synth'
        ]);
        // Assert that there was no providerError in CDK's stderr
        // Because we rely on the app/framework to actually error in case the
        // provider fails, we inspect the logs here.
        expect(output).not.toContain('$providerError');
    }
});
test_helpers_1.integTest('generating and loading assembly', async () => {
    const asmOutputDir = path.join(os.tmpdir(), 'cdk-integ-asm');
    await cdk_helpers_1.shell(['rm', '-rf', asmOutputDir]);
    // Make sure our fixture directory is clean
    await cdk_helpers_1.prepareAppFixture();
    // Synthesize a Cloud Assembly tothe default directory (cdk.out) and a specific directory.
    await cdk_helpers_1.cdk(['synth']);
    await cdk_helpers_1.cdk(['synth', '--output', asmOutputDir]);
    // cdk.out in the current directory and the indicated --output should be the same
    await cdk_helpers_1.shell(['diff', 'cdk.out', asmOutputDir], {
        cwd: cdk_helpers_1.INTEG_TEST_DIR,
    });
    // Check that we can 'ls' the synthesized asm.
    // Change to some random directory to make sure we're not accidentally loading cdk.json
    const list = await cdk_helpers_1.cdk(['--app', asmOutputDir, 'ls'], { cwd: os.tmpdir() });
    // Same stacks we know are in the app
    expect(list).toContain(`${cdk_helpers_1.STACK_NAME_PREFIX}-lambda`);
    expect(list).toContain(`${cdk_helpers_1.STACK_NAME_PREFIX}-test-1`);
    expect(list).toContain(`${cdk_helpers_1.STACK_NAME_PREFIX}-test-2`);
    // Check that we can use '.' and just synth ,the generated asm
    const stackTemplate = await cdk_helpers_1.cdk(['--app', '.', 'synth', cdk_helpers_1.fullStackName('test-2')], {
        cwd: asmOutputDir,
    });
    expect(stackTemplate).toContain('topic152D84A37');
    // Deploy a Lambda from the copied asm
    await cdk_helpers_1.cdkDeploy('lambda', { options: ['-a', '.'], cwd: asmOutputDir });
    // Remove the original custom docker file that was used during synth.
    // this verifies that the assemly has a copy of it and that the manifest uses
    // relative paths to reference to it.
    await fs_1.promises.unlink(path.join(cdk_helpers_1.INTEG_TEST_DIR, 'docker', 'Dockerfile.Custom'));
    // deploy a docker image with custom file without synth (uses assets)
    await cdk_helpers_1.cdkDeploy('docker-with-custom-file', { options: ['-a', '.'], cwd: asmOutputDir });
});
async function listChildren(parent, pred) {
    const ret = new Array();
    for (const child of await fs_1.promises.readdir(parent, { encoding: 'utf-8' })) {
        const fullPath = path.join(parent, child.toString());
        if (await pred(fullPath)) {
            ret.push(fullPath);
        }
    }
    return ret;
}
async function listChildDirs(parent) {
    return listChildren(parent, async (fullPath) => (await fs_1.promises.stat(fullPath)).isDirectory());
}
