// @aws-cdk/aws-cloudwatch/test/test.composite-alarm.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'test alarm rule expression builder'(test) {
        const stack = new core_1.Stack();
        const testMetric = new lib_1.Metric({
            namespace: 'CDK/Test',
            metricName: 'Metric',
        });
        const alarm1 = new lib_1.Alarm(stack, 'Alarm1', {
            metric: testMetric,
            threshold: 100,
            evaluationPeriods: 3,
        });
        const alarm2 = new lib_1.Alarm(stack, 'Alarm2', {
            metric: testMetric,
            threshold: 1000,
            evaluationPeriods: 3,
        });
        const alarm3 = new lib_1.Alarm(stack, 'Alarm3', {
            metric: testMetric,
            threshold: 10000,
            evaluationPeriods: 3,
        });
        const alarm4 = new lib_1.Alarm(stack, 'Alarm4', {
            metric: testMetric,
            threshold: 100000,
            evaluationPeriods: 3,
        });
        const alarmRule = lib_1.AlarmRule.anyOf(lib_1.AlarmRule.allOf(lib_1.AlarmRule.anyOf(alarm1, lib_1.AlarmRule.fromAlarm(alarm2, lib_1.AlarmState.OK), alarm3), lib_1.AlarmRule.not(lib_1.AlarmRule.fromAlarm(alarm4, lib_1.AlarmState.INSUFFICIENT_DATA))), lib_1.AlarmRule.fromBoolean(false));
        new lib_1.CompositeAlarm(stack, 'CompositeAlarm', {
            alarmRule,
        });
        assert_1.expect(stack).to(assert_1.haveResource('AWS::CloudWatch::CompositeAlarm', {
            AlarmName: 'CompositeAlarm',
            AlarmRule: {
                'Fn::Join': [
                    '',
                    [
                        '(((ALARM(',
                        {
                            'Fn::GetAtt': [
                                'Alarm1F9009D71',
                                'Arn',
                            ],
                        },
                        ') OR OK(',
                        {
                            'Fn::GetAtt': [
                                'Alarm2A7122E13',
                                'Arn',
                            ],
                        },
                        ') OR ALARM(',
                        {
                            'Fn::GetAtt': [
                                'Alarm32341D8D9',
                                'Arn',
                            ],
                        },
                        ')) AND (NOT (INSUFFICIENT_DATA(',
                        {
                            'Fn::GetAtt': [
                                'Alarm4671832C8',
                                'Arn',
                            ],
                        },
                        ')))) OR FALSE)',
                    ],
                ],
            },
        }));
        test.done();
    },
};
